phoenix-core/src/main/java/org/apache/phoenix/schema/task/Task.java

```diff
@@ -75,6 +87,55 @@ public Void run() throws Exception {
         }
     }
 
+    private static List<Mutation> getMutationsForSystemTaskTable(
+            PhoenixConnection conn, PreparedStatement stmt,
+            boolean accessCheckEnabled) throws IOException {
+        // we need to mutate SYSTEM.TASK with HBase/login user if access is enabled.
+        if (accessCheckEnabled) {
+            return User.runAsLoginUser(() -> {
+                final RpcCall rpcContext = RpcUtil.getRpcContext();
+                // setting RPC context as null so that user can be reset
+                try {
+                    RpcUtil.setRpcContext(null);
+                    stmt.execute();
+                    // retrieve mutations for SYSTEM.TASK upsert query
+                    Iterator<Pair<byte[], List<Mutation>>> iterator =
+                        conn.getMutationState().toMutations();
+                    List<Mutation> taskMutations = iterator.next().getSecond();
+                    // we are expecting conn to be used for single upsert
+                    // query on SYSTEM.TASK
+                    if (iterator.hasNext()) {
+                        throw new IOException("Provided connection should only "
+                            + "contain mutations related to SYSTEM.TASK table");
+                    }
+                    return taskMutations;
+                } catch (SQLException e) {
+                    throw new IOException(e);
+                } finally {
+                    // setting RPC context back to original context of the RPC
+                    RpcUtil.setRpcContext(rpcContext);
+                }
+            });
+        } else {
+            try {
+                stmt.execute();
```

Comment:
Can we extract the this part into a small helper method? The `else` case is similar to the `if` just that we aren't running with impersonation